### PR TITLE
QuarkusComponentTest blogpost

### PR DESCRIPTION
**Pull request by mkouba**

### CODE REVIEW

File: _posts/2023-07-10-quarkus-component-test.adoc [hunk 1]
```diff
@@ -0,0 +1,277 @@
+---
+layout: post
+title: 'Explore a new way of testing CDI components in Quarkus'
+date: 2023-07-10
+tags: testing
+synopsis: 'Quarkus 3.2 introduced an experimental feature to ease the testing of CDI components and mocking of their dependencies.'
+author: mkouba
+---
+
+The Quarkus component model is built on top CDI.
```
Comment (gastaldi) [inline _posts/2023-07-10-quarkus-component-test.adoc:10]:
```suggestion
The Quarkus component model is built on top of CDI.
```

File: _posts/2023-07-10-quarkus-component-test.adoc [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+However, writing unit tests for beans without a running CDI container is often a tedious work.
+Without the container services up and running, all the work has to be done manually.
+First of all, no dependency injection is perfomed.
```
Comment (gastaldi) [inline _posts/2023-07-10-quarkus-component-test.adoc:13]:
```suggestion
First of all, no dependency injection is performed. 
```

File: _posts/2023-07-10-quarkus-component-test.adoc [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+synopsis: 'Quarkus 3.2 introduced an experimental feature to ease the testing of CDI components and mocking of their dependencies.'
+author: mkouba
+---
+
+The Quarkus component model is built on top CDI.
+However, writing unit tests for beans without a running CDI container is often a tedious work.
+Without the container services up and running, all the work has to be done manually.
+First of all, no dependency injection is perfomed. 
+Furthermore, no events are fired and no observers are notified.
+Also interceptors are not applied.
```
Comment (gastaldi) [inline _posts/2023-07-10-quarkus-component-test.adoc:15]:
```suggestion
Also, interceptors are not applied.
```

File: _posts/2023-07-10-quarkus-component-test.adoc [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+    public String ping() { <3>
+        return bar ? charlie.ping() : "nok";
+    }
+}
+----
+<1>	`Foo` depends on `Charlie` which declares a method `ping()`.
+<2>	`Foo` depends on the config property `bar`.
+<3>  The goal is to test this method which makes use of the `Charlie` dependency and the `bar` config property.
+
+Then, a simple component test looks like:
```
Comment (gastaldi) [inline _posts/2023-07-10-quarkus-component-test.adoc:76]:
```suggestion
Then, a simple component test looks like this:
```

File: _posts/2023-07-10-quarkus-component-test.adoc [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+It also registers a dedicated configuration object.
+The container is then stopped and the config is released during the `after all` test phase.
+The fields annotated with `@Inject` and `@InjectMock` are injected after a test instance is created.
+Finally, the CDI request context is activated and terminated per each test method.
+
+NOTE: By default, a new test instance is created for each test method. Therefore, a new CDI container is started for each test method. However, if the test class is annotated with `@org.junit.jupiter.api.TestInstance` and the test instance lifecycle is set to `org.junit.jupiter.api.TestInstance.Lifecycle.PER_CLASS` then the CDI container will be shared across all test method executions of a given test class. 
+
+=== Tested components
+
+By default, the types of all fields annotated with `@Inject` are considered the component types.
```
Comment (gastaldi) [inline _posts/2023-07-10-quarkus-component-test.adoc:129]:
```suggestion
By default, the types of all fields annotated with `@Inject` are considered component types.
```

File: _posts/2023-07-10-quarkus-component-test.adoc [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+However, you can also specify additional test components: either with the `@QuarkusComponentTest#value()` or programatically as the arguments of the <<advanced_feautures,`QuarkusComponentTestExtension(Class<?>...)`>> constructor.
```
Comment (gastaldi) [inline _posts/2023-07-10-quarkus-component-test.adoc:130]:
```suggestion
However, you can also specify additional test components: either with the `@QuarkusComponentTest#value()` or programmatically as the arguments of the <<advanced_features,`QuarkusComponentTestExtension(Class<?>...)`>> constructor. 
```

File: _posts/2023-07-10-quarkus-component-test.adoc [hunk 7]
```diff
@@ -0,0 +1,277 @@
...
+And the injected reference is an unconfigured Mockito mock.
+This mock can be injected in the test with `@io.quarkus.test.InjectMock` and configured with the Mockito API.
+
+=== Configuration
+
+A dedicated `SmallRyeConfig` is registered during the `before all` test phase.
+It’s possible to set the configuration properties with the `@TestConfigProperty` annotation or programatically with the `QuarkusComponentTestExtension#configProperty(String, String)` method.
+If you need to use the default values for missing config properties, then `@QuarkusComponentTest#useDefaultConfigProperties()` and `QuarkusComponentTestExtension#useDefaultConfigProperties()` might come in useful.
+
+[[advanced_feautures]]
```
Comment (gastaldi) [inline _posts/2023-07-10-quarkus-component-test.adoc:149]:
```suggestion
[[advanced_features]]
```

File: _posts/2023-07-10-quarkus-component-test.adoc [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+
+By default, the types of all fields annotated with `@Inject` are considered the component types.
+However, you can also specify additional test components: either with the `@QuarkusComponentTest#value()` or programatically as the arguments of the <<advanced_feautures,`QuarkusComponentTestExtension(Class<?>...)`>> constructor. 
+Finally, the static nested classes declared on the test class are components too.
+
+TIP: Static nested classed declared on a test class that is annotated with `@QuarkusComponentTest` are excluded from bean discovery when running a regular `@QuarkusTest` in order to prevent unintentional CDI conflicts.
+
+=== Automatic mocking of unsatisfied dependencies
+
+Unlike in regular CDI environments the test does not fail if a component injects an unsatisfied dependency.
```
Comment (gastaldi) [inline _posts/2023-07-10-quarkus-component-test.adoc:137]:
```suggestion
Unlike in regular CDI environments, the test does not fail if a component injects an unsatisfied dependency.
```

File: _posts/2023-07-10-quarkus-component-test.adoc [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+Instead, a mock bean is registered automatically for each combination of required type and qualifiers of an injection point that resolves to an unsatisfied dependency.
+The mock bean has the `@Singleton` scope so it’s shared across all injection points with the same required type and qualifiers.
+And the injected reference is an unconfigured Mockito mock.
+This mock can be injected in the test with `@io.quarkus.test.InjectMock` and configured with the Mockito API.
+
+=== Configuration
+
+A dedicated `SmallRyeConfig` is registered during the `before all` test phase.
+It’s possible to set the configuration properties with the `@TestConfigProperty` annotation or programatically with the `QuarkusComponentTestExtension#configProperty(String, String)` method.
```
Comment (gastaldi) [inline _posts/2023-07-10-quarkus-component-test.adoc:146]:
```suggestion
It’s possible to set the configuration properties with the `@TestConfigProperty` annotation or programmatically with the `QuarkusComponentTestExtension#configProperty(String, String)` method.
```

File: _posts/2023-07-10-quarkus-component-test.adoc [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+    @Test
+    public void testPing() {
+        Mockito.when(charlieMock.ping()).thenReturn("OK");
+        assertEquals("OK", foo.ping());
+    }
+}
+----
+<1> Annotate a `static` field of type `QuarkusComponentTestExtension` with the `@RegisterExtension` annotation and configure the extension programatically.
+
+Sometimes you need the full control over the bean attributes and maybe even configure the default behavior of a mocked dependency.
```
Comment (gastaldi) [inline _posts/2023-07-10-quarkus-component-test.adoc:186]:
```suggestion
Sometimes you need full control over the bean attributes and maybe even configure the default behavior of a mocked dependency.
```

File: _posts/2023-07-10-quarkus-component-test.adoc [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+    Charlie charlieMock;
+
+    @Test
+    public void testPing() {
+        Mockito.when(charlieMock.ping()).thenReturn("OK");
+        assertEquals("OK", foo.ping());
+    }
+}
+----
+<1> Annotate a `static` field of type `QuarkusComponentTestExtension` with the `@RegisterExtension` annotation and configure the extension programatically.
```
Comment (gastaldi) [inline _posts/2023-07-10-quarkus-component-test.adoc:184]:
```suggestion
<1> Annotate a `static` field of type `QuarkusComponentTestExtension` with the `@RegisterExtension` annotation and configure the extension programmatically.
```

File: _posts/2023-07-10-quarkus-component-test.adoc [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+----
+<1> The scope of the mocked bean is `@Dependent`.
+<2> Configure the default behavior of the mock.
+
+=== Mocking CDI interceptors
+
+NOTE: This feature is only available in Quarkus 3.3+.
+
+If a tested component class declares an interceptor binding then you might need to mock the interception too.
+You can define an mock interceptor class as a static nested class of the test class.
```
Comment (gastaldi) [inline _posts/2023-07-10-quarkus-component-test.adoc:228]:
```suggestion
You can define a mock interceptor class as a static nested class of the test class.
```

File: _posts/2023-07-10-quarkus-component-test.adoc [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+import io.quarkus.test.InjectMock;
+import io.quarkus.test.component.QuarkusComponentTestExtension;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+
+public class FooTest {
+
+    @RegisterExtension 
+    static final QuarkusComponentTestExtension extension = new QuarkusComponentTestExtension()
+            .configProperty("bar","true");
```
Comment (Ladicek) [inline _posts/2023-07-10-quarkus-component-test.adoc:204]:
```suggestion
            .configProperty("bar","true")
```

**Comment by Ladicek** [comment]
Nice! The only thing I don't really like is how `QuarkusComponentTestExtension` is mentioned a few times _before_ the section "Advanced features", which makes it a bit harder to comprehend.

**Review by gastaldi — APPROVED** [review]

**Comment by mkouba** [comment]
> Nice! The only thing I don't really like is how `QuarkusComponentTestExtension` is mentioned a few times _before_ the section "Advanced features", which makes it a bit harder to comprehend.

Yeah, I know. I've modified the example code so that it's clear that the `@QuarkusComponentTest`  registers the `QuarkusComponentTestExtension`... it's not optimal but hopefully a little bit better.

**Comment by mkouba** [comment]
Thanks @gastaldi for all your comments! :+1:

**Review by Ladicek — APPROVED** [review]

**Pull request merged**